assembly/src/parsers/mod.rs

```diff
@@ -1,21 +1,419 @@
-use super::{AssemblyError, CodeBlock, Token, TokenStream};
-pub use blocks::combine_blocks;
-use vm_core::{utils::collections::Vec, Felt, Operation, StarkField};
+use super::{
+    errors::SerializationError, BTreeMap, Felt, ParsingError, ProcedureId, StarkField, String,
+    ToString, Token, TokenStream, Vec, MODULE_PATH_DELIM,
+};
+use core::{fmt::Display, ops::Deref};
+use serde::{ByteReader, ByteWriter, Deserializable, Serializable};
 
-mod blocks;
+mod nodes;
+pub(crate) use nodes::{Instruction, Node};
 
-mod ast;
-pub use ast::{parse_module, ModuleAst, NamedModuleAst, ProcedureAst};
-pub(crate) use ast::{parse_program, Instruction, Node, ProgramAst};
+mod context;
+use context::ParserContext;
+
+mod field_ops;
+mod io_ops;
+mod serde;
+mod stack_ops;
+mod u32_ops;
+
+#[cfg(test)]
+pub mod tests;
+
+// TYPE ALIASES
+// ================================================================================================
+type LocalProcMap = BTreeMap<String, (u16, ProcedureAst)>;
+
+// ABSTRACT SYNTAX TREE STRUCTS
+// ================================================================================================
+
+/// An abstract syntax tree (AST) of a Miden program.
+///
+/// A program AST consists of a list of internal procedure ASTs and a list of body nodes.
+#[derive(Debug, Eq, PartialEq)]
+pub struct ProgramAst {
+    pub local_procs: Vec<ProcedureAst>,
+    pub body: Vec<Node>,
+}
+
+#[allow(dead_code)]
```

Comment:
Should we keep this here?

Reply:
Good catch! We should actually export `ProgramAst` publicly. Will update in a follow-up PR.